src/Instrumentation/Doctrine/src/DoctrineInstrumentation.php

```diff
@@ -0,0 +1,182 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\Doctrine;
+
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\API\Trace\SpanBuilderInterface;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\API\Trace\StatusCode;
+use OpenTelemetry\Context\Context;
+
+use function OpenTelemetry\Instrumentation\hook;
+
+use OpenTelemetry\SemConv\TraceAttributes;
+use Throwable;
+
+class DoctrineInstrumentation
+{
+    public const NAME = 'doctrine';
+
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.doctrine');
+
+        hook(
+            \Doctrine\DBAL\Driver::class,
+            'connect',
+            pre: static function (\Doctrine\DBAL\Driver $driver, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'Doctrine\DBAL\Driver::connect', $function, $class, $filename, $lineno)
+                    ->setSpanKind(SpanKind::KIND_CLIENT);
+                $builder
+                    ->setAttribute(TraceAttributes::SERVER_ADDRESS, $params[0]['host'] ?? 'unknown')
+                    ->setAttribute(TraceAttributes::SERVER_PORT, $params[0]['port'] ?? 'unknown')
+                    ->setAttribute(TraceAttributes::DB_SYSTEM, $params[0]['driver'] ?? 'unknown')
+                    ->setAttribute(TraceAttributes::DB_NAMESPACE, $params[0]['dbname'] ?? 'unknown');
+                $parent = Context::getCurrent();
+                $span = $builder->startSpan();
+                Context::storage()->attach($span->storeInContext($parent));
+            },
+            post: static function (\Doctrine\DBAL\Driver $driver, array $params, ?\Doctrine\DBAL\Driver\Connection $connection, ?Throwable $exception) {
+                self::end($exception);
+            }
+        );
+
+        hook(
+            \Doctrine\DBAL\Driver\Connection::class,
+            'query',
+            pre: static function (\Doctrine\DBAL\Driver\Connection $connection, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = self::makeBuilder($instrumentation, 'Doctrine\DBAL\Driver\Connection::query', $function, $class, $filename, $lineno)
```

Comment:
The recommended span name here would be something like `SELECT my_table`.

The duplication with the pdo instrumentation is the same here.

https://opentelemetry.io/docs/specs/semconv/database/database-spans/#name

Reply:
Ok. Do you know a good agnostic SQL parser to achieve this? The [one I found ](https://github.com/phpmyadmin/sql-parser) is focusing on Mysql syntax.

Reply:
Doctrine already knows the table name(s).

Reply:
> Doctrine already knows the table name(s).

With the approach I'm following right now, I dont have this information. Are you referring to Doctrine ORM? I'm creating hooks on the methods of `\Doctrine\DBAL\Driver` and `\Doctrine\DBAL\Driver\Connection`

Reply:
I see, that's a good point.

The issue with this approach is whichever parser you choose is bound to be quite slow and will probably be slowest part of the call (which includes the actual query), making it unusable for production use. ORM itself has a DQL query parser (which is close enough for high level comparison) and it's basically unusable in prod without it being cached with [ORM query cache](https://www.doctrine-project.org/projects/doctrine-orm/en/3.3/reference/caching.html#query-cache).

Running a random query parser on every DB query is unlikely to be viable.

Reply:
IMO we could consider the ORM instrumentation even if it's still not here yet.

For example, since ORM will know what the table names are, you could assume they will be put in context by the ORM instrumentation and then only try to detect it if not there? This would future proof this implementation and allow extensions when they happen.

Reply:
Sorry for the late response but I'm pretty busy at the moment. 



> IMO we could consider the ORM instrumentation even if it's still not here yet.
> 
> For example, since ORM will know what the table names are, you could assume they will be put in context by the ORM instrumentation and then only try to detect it if not there? This would future proof this implementation and allow extensions when they happen.

I see your point but my focus is to implement hooks for Doctrine DBAL and in the future time permitted we could refactor the functions to behave as you suggested.